prow/tide/tide.go

```diff
@@ -573,6 +581,90 @@ func filterPR(ghc githubClient, sp *subpool, pr *PullRequest) bool {
 	return false
 }
 
+// mergeMethodChecker provides a function to check if a PR can be merged with
+// the requested method and does not have a merge conflict.
+// It caches results and should be cleared periodically with clearCache()
+type mergeMethodChecker struct {
+	config config.Getter
+	ghc    githubClient
+	log    *logrus.Entry
+
+	sync.Mutex
+	cache map[config.OrgRepo]map[github.PullRequestMergeType]bool
+}
+
+func newMergeMethodChecker(cfg config.Getter, ghc githubClient, log *logrus.Entry) *mergeMethodChecker {
+	m := &mergeMethodChecker{
+		config: cfg,
+		ghc:    ghc,
+		log:    log,
+		cache:  map[config.OrgRepo]map[github.PullRequestMergeType]bool{},
+	}
+
+	go m.clearCache()
+	return m
+}
+
+func (m *mergeMethodChecker) clearCache() {
+	// Only do this once per token reset since it could be a bit expensive for
+	// Tide instances that handle hundreds of repos.
+	ticker := time.NewTicker(time.Hour)
+	for {
+		<-ticker.C
+		m.Lock()
+		m.cache = make(map[config.OrgRepo]map[github.PullRequestMergeType]bool)
+		m.Unlock()
+	}
+}
+
+func (m *mergeMethodChecker) repoMethods(orgRepo config.OrgRepo) (map[github.PullRequestMergeType]bool, error) {
+	m.Lock()
+	defer m.Unlock()
+
+	repoMethods, ok := m.cache[orgRepo]
+	if !ok {
+		fullRepo, err := m.ghc.GetRepo(orgRepo.Org, orgRepo.Repo)
+		if err != nil {
+			return nil, err
+		}
+		repoMethods = map[github.PullRequestMergeType]bool{
+			github.MergeMerge:  fullRepo.AllowMergeCommit,
+			github.MergeSquash: fullRepo.AllowSquashMerge,
+			github.MergeRebase: fullRepo.AllowRebaseMerge,
+		}
+		m.cache[orgRepo] = repoMethods
+	}
+	return repoMethods, nil
+}
+
+func (m *mergeMethodChecker) isAllowed(pr *PullRequest) error {
```

Comment:
Lets please change the return of this to properly distinguish "is merge method allowed" from "an error occurred when checking". The caller can then still decide to drop the PR if either of the two is not as expected.